contrib/pg_tde/src/access/pg_tde_tdemap.c

```diff
@@ -76,50 +76,52 @@ typedef struct RelKeyCacheRec
  *
- * This is a slice backed by memory `*data`. Initially, we allocate one memory
- * page (usually 4Kb). We reallocate it by adding another page when we run out
- * of space. This memory is locked in the RAM so it won't be paged to the swap
- * (we don't want decrypted keys on disk). We do allocations in mem pages as
- * these are the units `mlock()` operations are performed in.
+ * This is a slice backed by memory `*data`. This memory is locked into RAM so
+ * it won't be paged to the swap (we don't want decrypted keys on disk). We do
+ * allocations in mem pages as these are the units `mlock()` operations are
+ * performed in.
  *
  * Currently, the cache can only grow (no eviction). The data is located in
  * TopMemoryContext hence being wiped when the process exits, as well as memory
  * is being unlocked by OS.
  */
-typedef struct RelKeyCache
+typedef struct LockedMemory
 {
-	RelKeyCacheRec *data;		/* must be a multiple of a memory page
+	unsigned char *data;		/* must be a multiple of a memory page
 								 * (usually 4Kb) */
-	int			len;			/* num of RelKeyCacheRecs currenty in cache */
-	int			cap;			/* max amount of RelKeyCacheRec data can fit */
-} RelKeyCache;
+	int			len;
+	int			cap;
+} LockedMemory;
 
-RelKeyCache tde_rel_key_cache = {
+
+LockedMemory tde_safe_cache = {
 	.data = NULL,
 	.len = 0,
 	.cap = 0,
 };
 
-
 /*
  * TODO: WAL should have its own RelKeyCache
  */
 static WALKeyCacheRec *tde_wal_key_cache = NULL;
 static WALKeyCacheRec *tde_wal_key_last_rec = NULL;
 
-static InternalKey *pg_tde_get_key_from_file(const RelFileLocator *rlocator, uint32 key_type);
+static bool pg_tde_get_key_from_file(const RelFileLocator *rlocator, uint32 key_type, InternalKey *out);
 static TDEMapEntry *pg_tde_find_map_entry(const RelFileLocator *rlocator, uint32 key_type, char *db_map_path);
-static InternalKey *tde_decrypt_rel_key(TDEPrincipalKey *principal_key, TDEMapEntry *map_entry);
+static void tde_decrypt_rel_key(TDEPrincipalKey *principal_key, TDEMapEntry *map_entry, InternalKey *out);
 static int	pg_tde_open_file_basic(const char *tde_filename, int fileFlags, bool ignore_missing);
 static void pg_tde_file_header_read(const char *tde_filename, int fd, TDEFileHeader *fheader, off_t *bytes_read);
 static bool pg_tde_read_one_map_entry(int fd, const RelFileLocator *rlocator, int flags, TDEMapEntry *map_entry, off_t *offset);
 static TDEMapEntry *pg_tde_read_one_map_entry2(int keydata_fd, int32 key_index, TDEPrincipalKey *principal_key);
 static int	pg_tde_open_file_read(const char *tde_filename, off_t *curr_pos);
 static InternalKey *pg_tde_get_key_from_cache(const RelFileLocator *rlocator, uint32 key_type);
 static WALKeyCacheRec *pg_tde_add_wal_key_to_cache(InternalKey *cached_key, XLogRecPtr start_lsn);
-static InternalKey *pg_tde_put_key_into_cache(const RelFileLocator *locator, InternalKey *key);
+static void *tde_alloc_safe_mem(size_t len);
+static inline void tde_release_safe_mem(size_t len);
```

Comment:
I am not the biggest fan of explicit inline unless there is a good reason for it. I think the optimizer would inline this when it should anyway.